### PR TITLE
Add option to run examples directly with binder

### PR DESCRIPTION
**Pull request by naveen521kk**
## Overview: What does this pull request change?

There's now a "Run" button at the bottom of the code examples which allows running the examples directly using binder.

https://github.com/ManimCommunity/manim/assets/49693820/0ff9890b-84aa-4f65-b1fb-2f4acfe8618f

## Motivation and Explanation: Why and how do your changes improve the library?

This should allow people to try manim directly from the documentation. Inspiration is from https://github.com/ines/juniper.

## Links to added or changed documentation pages

https://manimce--3427.org.readthedocs.build/en/3427/

## Further Information and Comments

The minified JS is from https://github.com/naveen521kk/manim-binder. Should I move that repository to ManimCommunity?

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

**Comment by naveen521kk** [comment]
> When I try with videos I can't seem to get it working; the video does not load correctly:

What browser are you using? I've tested all this in Firefox and everything seems to work there. I'll also check with chrome and see whether it works there as well.

**Comment by naveen521kk** [comment]
Just so you know, currently, only the recent version of Manim is run, ie. run the Jupyter environment from https://github.com/ManimCommunity/jupyter_examples Dockerfile. Since this is just added this won't be a problem but in the future when new versions with breaking changes are released these old examples may not work. 

It would be better if we could make a specific version of the documentation point to the specific Jupyter environment so that old examples will run as expected. If we are to do that then we would need corresponding branches for these in [jupyter_examples](https://github.com/ManimCommunity/jupyter_examples) repo. Also, while releasing a new version it would be necessary to create a branch/tag on the other repository too. Also, since RTD provides docs of the main branch, it would be nice to have that too.

What do you guys think about this? Is it worth the effort?

File: manim/utils/docbuild/manim_directive.py [hunk 1]
```diff
@@ -231,10 +239,15 @@ def run(self):
...
         source_block = [
-            ".. code-block:: python",
+            ".. raw:: html",
             "",
+            "    <pre data-interactive>",
             "    from manim import *\n",
             *("    " + line for line in self.content),
+            "",
+            "    # don't remove below command for run button to work",
+            f"    %manim -qm {clsname}",
```
Comment (behackl) [inline manim/utils/docbuild/manim_directive.py:249]: I don't necessarily like that these lines are appended to every example in the documentation, but I'd be wiilling to go with it for an initial version. I'll leave some general comments in the overall review.

**Review by behackl — COMMENTED** [review]
The functionality is really cool and useful -- and definitely also means that we should revisit the content in our example gallery. The order of examples, and how they are grouped together makes limited sense, but well, that's a different project altogether.

Here is just a suggestion for a slightly different approach that would require some changes to the architecture, but they would (I think at least!) not be entirely unreasonable:
- We keep the default rendering of the example boxes, and just render an additional button ("Make interactive" or so?) between the code and the references. We can add, say, the `clsname` also in a `data-manim-classname`-attribute or so, so that it is accessible in the js too.
- The button, once clicked, basically calls your `renderCell` function for that one particular code block. The first time this happens, the binder connection is established too. (UX wise, we could just disable the button and show a spinner.)
- If the connection to binder is established, `renderCell` is actually executed, the codemirror editor is shown. This is also the point where I'd suggest injecting the newline, the comment about rendering, and the actual magic command.
- This also removes the "Make interactive"-button, and replaces it with the "Run"-button.

This preserves the current "feature" that by default, examples in our documentation can be copy-pasted (and I reckon we would get quite some bug reports because "code does not run on my machine"). I think designing the act of making the example interactive as a conscious decision is beneficial too, and users would interact with it in a somewhat more aware state. What do you think?

I should say explicitly: these are fairly high-level remarks. I am very eager to make this feature available for people; it really is pretty awesome! (And that's also why I think trying to make the user experience as smooth as possible with this would be worth it.)

**Comment by naveen521kk** [comment]
The method you suggested is a lot better than how it is currently. I'll work on this week, it'll take some time though. Do you want to merge this PR for `v0.18.0` or wait for it to get ready? Also, what do you think about https://github.com/ManimCommunity/manim/pull/3427#issuecomment-1793516930?

**Comment by naveen521kk** [comment]
I ended up implementing it 😄 (in a way which is easy to implement)

https://github.com/ManimCommunity/manim/assets/49693820/45a474a3-cac6-4d53-a6d0-31fbff689b61

Should I remove the original code block when the "Make interactive" button is clicked?

> The first time this happens, the binder connection is established too. (UX wise, we could just disable the button and show a spinner.)

Something which I didn't implement is a loading spinner. A spinner doesn't really look good and just the loading text would be good, IMO. Also, the binder connection is only established when the "Run" button is clicked, which do you think is better?

**Review by behackl — APPROVED** [review]
I think this is more than good enough to include it as is. Thank you very much for your efforts, I like this current version a lot!

**Comment by behackl** [comment]
> Just so you know, currently, only the recent version of Manim is run, ie. run the Jupyter environment from https://github.com/ManimCommunity/jupyter_examples Dockerfile. Since this is just added this won't be a problem but in the future when new versions with breaking changes are released these old examples may not work.
> 

I am not so much concerned about making sure that old versions of the documentation can be executed interactively too. The latest (or at least the stable version, which by far the most people use) should work, as we usually take care that docbuilds are successful anyways.

> It would be better if we could make a specific version of the documentation point to the specific Jupyter environment so that old examples will run as expected. If we are to do that then we would need corresponding branches for these in [jupyter_examples](https://github.com/ManimCommunity/jupyter_examples) repo. Also, while releasing a new version it would be necessary to create a branch/tag on the other repository too. Also, since RTD provides docs of the main branch, it would be nice to have that too.
> 
> What do you guys think about this? Is it worth the effort?

It would certainly be possible, sure. For now, I don't think it is worth it -- we can see what people think when we communicate this new documentation feature with the new release, and if this is something that is requested, we can come back to it. 👍

**Comment by naveen521kk** [comment]
Actually, after the refactor I ended up implementing version based builds, so now things build from `v0.17.3` branch in jupyter_examples. After releasing a new version we'd have to create a corresponding branch there for it to work like `v0.18.0`.

**Comment by naveen521kk** [comment]
Actually, after the refactor I ended up implementing version based builds, so now things build from `v0.17.3` branch in jupyter_examples. After releasing a new version we'd have to create a corresponding branch there for it to work like `v0.18.0`.

**Pull request merged**